### ISSUE TITLE
File ini

### ISSUE THREAD

**Issue by Krzychumag** [cutechess/cutechess#264]
Please add an option to read the ini file chess engine.

**Comment by alwey** [comment]
Perhaps this is already in cutechess, could you please explain the problem in more detail?

**Comment by Krzychumag** [comment]
Of course. Please select a chess engine EXchess 7.97b x64 and choose " Advanced ". As you can see there are only two options. If the engine read from the ini file then there would be more options. Na przykład mozna by było ustawić tablebase Nalimova. And unfortunately it can not.

**Comment by alwey** [comment]
Compiled from sources and installed EXchess 7.97 but I have no .ini file.
I only found 
```
White-To-Move[1]: xboard
protover 2

feature setboard=1
feature playother=1
feature usermove=1
feature ics=1
feature smp=1
feature memory=1
feature variants="normal,nocastle,fischerandom"
feature myname="EXchess v7.97b"
feature option="Playing Strength -slider 100 1 100
feature done=1


```
The feature option `Playing Strength` seems to be broken - the closing quotation mark is missing and cutechess does not like it. The error is in EXchess's  `main,cpp, line 798`.

@Krzychumag : Could you post an example .ini file for me?

**Comment by Krzychumag** [comment]
I guess I made a mistake, but the point is to open such a file.

[search.txt](https://github.com/cutechess/cutechess/files/1235087/search.txt)

**Comment by alwey** [comment]
Cutechess does not prevent engines from reading their files. In cutechess-cli you can transfer arguments to the engines and you can transfer init strings. From the manual page cutechess-cli.6, subarguments to `-engine`:

```
     arg=arg
             Pass arg to the engine as a command line argument.

     initstr=arg
             Send arg to the engine's standard input at `startup.`

```

**Comment by alwey** [comment]
In the GUI it is similar: Amend the call of the engine binary with command line arguments, and if necessary enter init strings into the text box below.

**Comment by Krzychumag** [comment]
For me personally this is a program bug. The same applies to the engine Scorpio 2.7.7. I'm sorry to write so but I think to change it. I just do not know how to program.

**Comment by alwey** [comment]
I can not read the search.txt file - what is in it?

**Comment by alwey** [comment]
> For me personally this is a program bug.
I think we can agree that cutechess is less buggy than most engines?

**Comment by Krzychumag** [comment]
# This file contains the search and learning paramaters for EXchess.
#
# You are welcome to change these parameters to modify EXchess'
# style of play.  The default values are the best I've found
# so far. 
#
# EXchess will look for this file in the same directory as the
# executable.  (Under normal circumstances it will be there)
#
# If this file is not found by EXchess, the default (pre-programmed)
# parameters will be used.
#

# Hash Tables - larger is generally better,
#               but don't exceed your available RAM, if you
#               do - the search will slow down to a crawl.

HASH  1024  

# For some programs using the xboard protocol, set XBOARD to 1
# This will make EXchess start in xboard mode, rather than waiting
# for the "xboard" command from the host program.  Only use this
# if necessary

XBOARD  0

# Define default number of threads or processor 'cores' to use 
#  --> do not exceed the number of processors in your machine!

THREADS  4 
           # by the 'cores' command, either on the
           # command line or while the program is running
	   # Maximum value = 32  (hard coded limit)

# Endgame Tablebases - if you have them installed

EGTB_PATH D:\Nalimov
EGTB_CACHE_SIZE 32.0 

# Log File

LOG  0  # set to 0 to turn off, 1 for logging on
MAX_LOGS 100  # maximum number of log files, 1 = append to single file 
	      #                             -1 = overwrite single file

# Opening book stuff

BOOK_FILE      main_bk.dat      # Name of the main opening book file

# Chess Skill to weaken play
CHESS_SKILL        100          # Set between 1 and 100 for percent of skill used 
                                #  -- values less than 50 will begin to cause oversights
                                #     in addition to reducing knowledge used    

# Null move
#
# Replacing NULL_MOVE with 0 will turn off Null move 

NULL_MOVE            1

**Comment by alwey** [comment]
@Krzychumag : Please encapsulate the text with the code markers (Symbol <> above the edit box).

**Comment by Krzychumag** [comment]
`# This file contains the search and learning paramaters for EXchess.`
`#`
`# You are welcome to change these parameters to modify EXchess'`
`# style of play.  The default values are the best I've found`
`# so far. `
`#`
`# EXchess will look for this file in the same directory as the`
`# executable.  (Under normal circumstances it will be there)`
`#`
`# If this file is not found by EXchess, the default (pre-programmed)`
`# parameters will be used.`
`#`

`# Hash Tables - larger is generally better,`
`#               but don't exceed your available RAM, if you`
`#               do - the search will slow down to a crawl.`

`HASH  1024  `

`# For some programs using the xboard protocol, set XBOARD to 1`
`# This will make EXchess start in xboard mode, rather than waiting`
`# for the "xboard" command from the host program.  Only use this`
`# if necessary`

`XBOARD  0`

`# Define default number of threads or processor 'cores' to use `
`#  --> do not exceed the number of processors in your machine!`

`THREADS  4 `
          ` # by the 'cores' command, either on the`
           `# command line or while the program is running`
	 `  # Maximum value = 32  (hard coded limit)`

`# Endgame Tablebases - if you have them installed`

`EGTB_PATH D:\Nalimov`
`EGTB_CACHE_SIZE 32.0 `

`# Log File`

`LOG  0  # set to 0 to turn off, 1 for logging on`
`MAX_LOGS 100  # maximum number of log files, 1 = append to single file `
	     ` #                             -1 = overwrite single file
`
`# Opening book stuff`

`BOOK_FILE      main_bk.dat      # Name of the main opening book file`

`# Chess Skill to weaken play`
`CHESS_SKILL        100          # Set between 1 and 100 for percent of skill used `
                               ` #  -- values less than 50 will begin to cause oversights`
                               ` #     in addition to reducing knowledge used    `

`# Null move`
`#`
`# Replacing NULL_MOVE with 0 will turn off Null move `

`NULL_MOVE            1        `

**Comment by Krzychumag** [comment]
When he enters " Advanced " we have only two options core and memory. And where are the other options I find. That is the question. No matter what I choose, whether uci or xboard. Unfortunately, but it is poorly done.

**Comment by alwey** [comment]
Repaired EXchess main.cpp, line 907 and renamed the settings file to `search.par`: Engine reads the file.
But ...
```

Hash size = 12582912 buckets of 4 entries, 768 Mbytes
Pawn size = 2013266 individual entries, 76 Mbytes
Score size = 7549747 individual entries, 115 Mbytes
Cmove size = 2097152 individual entries, 64 Mbytes

Type 'help' for a list of commands.
White-To-Move[1]: protover 2

feature setboard=1
feature playother=1
feature usermove=1
feature ics=1
feature smp=1
feature memory=1
feature variants="normal,nocastle,fischerandom"
feature myname="EXchess v7.97b"
feature option="Playing Strength -slider 100 1 100"
feature done=1

```

I do not see many options listed here.

**Comment by alwey** [comment]
Which UI shows you more options with this engine version?

**Comment by Krzychumag** [comment]
I do not know. I am using another GUI file search.par. I loaded the engine EXchess v7.97b and still in the tab " Advanced " are only two options core and memory. You can not set other options, such as Nalimova tablebase. Unfortunately it's not good that there is no going to set other options.  Try another engine that uses the ini file - crafty.

**Comment by alwey** [comment]
@Krzychumag : 
> Unfortunately, but it is poorly done.

Great remark. I encourage you to be more positive. You even might help the project by contributing a solution. 

As of now, from my point of view, this issue #264 is none.

**Comment by alwey** [comment]
@Krzychumag : This is an internal thing of the engine. Cutechess ist out of it.

**Comment by Krzychumag** [comment]
This is not an internal engine affair unfortunately. This is just a matter of program only Cutechess. Remember one important principle of life. How to do something to do it well. Another GUI shows a program Cutechess does not show. You need to make an entry in the program in order for the programs to use the ini and search. I will be more positive as everything will work properly.

Good night.

**Comment by alwey** [comment]
@Krzychumag :
> Another GUI shows a program Cutechess does not show

I already asked you - which GUI?

**Comment by alwey** [comment]
@ilaripih : I now feel a bit trolled ... I will not interfere with this issue anymore.

**Comment by Krzychumag** [comment]
I already asked you - which GUI? - ChessGui 0.245m - http://www.chess.hylogic.de/
I have a slightly different impression - that you can not cope with the problem I described. 53/5000
Make a test with the Crafty engine that uses the file ini.

**Comment by Krzychumag** [comment]
Please see the engine Crafty 25.0.2 x64. The program has only one option in the tab " Advanced "  - cores.  
In addition, nothing can be set - neither the hash nor the sysygy tablebases.

**Comment by alwey** [comment]
![a2](https://user-images.githubusercontent.com/6425738/29488806-9d36e232-8512-11e7-8b96-fb265d28eb17.png)

**Comment by Krzychumag** [comment]
@alwey 
You can write about what's going on. Because I do not quite understand.

**Comment by fsmosca** [comment]
ini or config files are common in xboard engines. Perhaps equip cutechess with a button to open a file with a text editor, something like the following. 

![image](https://user-images.githubusercontent.com/22366935/29495316-72dab77e-85ef-11e7-84ba-4b1ee891a8f2.png)

@Krzychumag 
CECP protocol had already provided **feature** option to standardize the engine settings communications with the gui, but it seems like engine authors does not implement everything, perhaps for some other reasons.

**Comment by Krzychumag** [comment]
Unfortunately, I do not have such a window like you. I only have talkie windows.
![1](https://user-images.githubusercontent.com/30792281/29496753-64957ba6-85d9-11e7-9bb0-af1d414d8303.jpg)
![2](https://user-images.githubusercontent.com/30792281/29496756-6b90bd62-85d9-11e7-9a10-e7083acaa59a.jpg)

**Comment by alwey** [comment]
@fsmosca : Thank you for clearing this up.

@Krzychumag , @fsmosca 
In my opinion this is going away from the core things cutechess should know about and more into the user's realm.

As you stated the CECP protocol (UCI has something similar) provides feature options. So the UI can provide means to display and/or to configure these options without knowing anything about the meaning of these options. This is what xboard, cutechess, Arena, ... you name it ..., support.

Now there are engines that are [additionally] configurable in other ways, by command line arguments, by initialisation commands, or by external configuration files. The first two of these methods are supported by cutechess. They are part of the communication process between cutechess, the operating system and the engine. The third method is essentially the engine reading its external configuration file(s). One problem: There is no general rule how the UI can identify these configuration files.

So the user has to specify and tell the UI, say Cutechess GUI, which files are relevant to configure the engine. This could be augmented by file selection methods and the information could be saved into the engine configuration sections of cutechess.. If the user is successful, cutechess will connect these files to an editing process. 

I admit, it is nice to have support for this, especially for those users who are uncomfortable with directly attacking configurations with an editor.

@Krzychumag , @fsmosca : +1 for this feature request

But I feel, there are some issues that might enjoy higher priority at the moment.

**Comment by Krzychumag** [comment]
I do not quite understand it all. **I think it should be the highest and most important priority.** Because I can not imagine such a situation that the program does not read all the engine options. Someone will just think it's a poor GUI.

**Comment by alwey** [comment]
@Krzychumag : 
>Someone will just think it's a poor GUI.

There is some misunderstanding. Cutechess originally is a tool for experts. The CLI since long has been the flagship of the project. The GUI is a complement and evolving, you can have best results when you use both tools, depending on the problem.

>Because I can not imagine such a situation that the program does not read all the engine options. 

Many interested users have tested these engines without having the newly requested feature and they are having full control. They are just reading manuals and readme files, using editors for the configuration files,  or even writing engines themselves. And usually the engines do read their specific configuration files with all options. They just do not reflect in the GUI. Example: I used EXchess successfully (see above) and after you had given me the `search.par` file I made changes to the current hash configuration using command line arguments (in the area of the green arrow above).

> I do not quite understand it all. I think it should be the highest and most important priority.

In my opinion issues like #250 (we essentially fixed this),  #231, #230, #211, #210 and #147 (made PR for the cli/lib part),  #213, #205, #188 (working on that), #116, #112, #109, #107, #106, #95 (working on that), #31 have a higher priority - and BTW - they are older.

For example, @fsmosca is patiently waiting for #213 without complaining or making waves, and #250 in my view is much more important than this issue.

**Comment by Krzychumag** [comment]
@alwey 
Sorry but you're writing nonsense. I have installed the program Cutechess 1.0. I want to test the engine
Crafty, Exchess, Scorpio. Please write why the engines do not show all the settings.

**Comment by alwey** [comment]
@Krzychumag : I am done here.

**Comment by Krzychumag** [comment]
@alwey 
Please answer the question.

**Comment by ianfab** [comment]
@Krzychumag 
Keep in mind that by downloading Cutechess you have **not** bought a commercial product with customer support. Nevertheless @alwey tried to help you, but you have been aggressively demanding answers/solutions, so I am not suprised that he has lost the motivation to help you solve your issue. As the question, so the answer.

**Comment by Krzychumag** [comment]
> Keep in mind that by downloading Cutechess you have not bought a commercial product with customer support. Nevertheless @alwey tried to help you, but you have been aggressively demanding answers/solutions, so I am not suprised that he has lost the motivation to help you solve your issue. As the question, so the answer.

Of course, I remember that it was not a commercial product. As for @alwey it must help everyone who uses the program CuteChess. This is not about my person. The problem with the chess engines I have written above has been around for a long time and so far, but unfortunately no one wants to solve it.

**Comment by fsmosca** [comment]
@Krzychumag 

> Krzychumag wrote: 
> Unfortunately, I do not have such a window like you. I only have talkie windows.

The window I showed, the one with **Open ini file** button is not real, I only edited the image.

**Comment by Krzychumag** [comment]
Thank you for the information fsmosca. A pity perhaps the problem would be solved.

**Comment by ilaripih** [comment]
I think it's a reasonable request that users should be able to open the configuration files with Notepad, their favorite editor etc. as long as the engine provides the option to choose the ini file (both Xboard and UCI make this possible). But this is not a high priority issue. Pull requests are always welcome though!

**Comment by Krzychumag** [comment]
> I think it's a reasonable request that users should be able to open the configuration files with Notepad, their favorite editor etc. as long as the engine provides the option to choose the ini file (both Xboard and UCI make this possible). But this is not a high priority issue. Pull requests are always welcome though!

This is a very serious mistake. Priority number 1 should be. 

@ilaripih 

Please do not write that it is not a bug since you can not set all options during the game.

**Comment by Krzychumag** [comment]
@ilaripih
Please test the chess engine Crafty 25.0.2 x64 and set in hash and tablebase syzygy. I'm sorry but this is not going to set. And you write that it's not a mistake.

**Comment by ilaripih** [comment]
When it comes to development of free open source software there are no serious mistakes. If I want to spend the next 2 months tweaking the piece animations or refactoring some already working code I can do that. That's a nice contrast to my daily job where other people set my priorities and schedules.

This is not a feature that's been requested by a lot of people so the priority is not the highest. We have issues that many people (including me) want desperately so those have higher priority. I agree that this would be a useful feature. On the other hand I want to push engine authors to fully adopt the chess protocol they're using: for example Crafty could easily use XBoard's custom options to set hash size and tablebase path in Cute Chess. In any other domain than computer chess the clients would have to adapt to the protocol and not the other way around. But in computer chess we unfortunately have things backwards and GUIs are forced to adapt to misbehaving clients. I've been trying to fight this mindset and have actually been somewhat successful doing it.

BUT: like I said, if anyone wants to dedicate their time to work on this issue or any other issue that makes sense, I'll gladly help them to understand our coding convention, refactor the code, suggest alternative ways to achieve something, etc. For example @alwey has implemented a lot of variants that I probably will never play and wouldn't have implemented myself, but I'm still really grateful for his contributions and I'm proud that Cute Chess has such good variant support.

**Comment by Krzychumag** [comment]
> When it comes to development of free open source software there are no serious mistakes. If I want to spend the next 2 months tweaking the piece animations or refactoring some already working code I can do that. That's a nice contrast to my daily job where other people set my priorities and schedules.

This is not a feature that's been requested by a lot of people so the priority is not the highest. We have issues that many people (including me) want desperately so those have higher priority. I agree that this would be a useful feature. On the other hand I want to push engine authors to fully adopt the chess protocol they're using: for example Crafty could easily use XBoard's custom options to set hash size and tablebase path in Cute Chess. In any other domain than computer chess the clients would have to adapt to the protocol and not the other way around. But in computer chess we unfortunately have things backwards and GUIs are forced to adapt to misbehaving clients. I've been trying to fight this mindset and have actually been somewhat successful doing it.

BUT: like I said, if anyone wants to dedicate their time to work on this issue or any other issue that makes sense, I'll gladly help them to understand our coding convention, refactor the code, suggest alternative ways to achieve something, etc. For example @alwey has implemented a lot of variants that I probably will never play and wouldn't have implemented myself, but I'm still really grateful for his contributions and I'm proud that Cute Chess has such good variant support.

@ilaripih
Unfortunately, but not this way. You have to do in the program to make chess engines can read ini file and other files. For now imagine what people are thinking of who are testing the Crafty chess engine about the program  Cute Chess. That in the program is a serious error since the program is not able to read the file ini and other files and can not set all the engine options.

**Comment by fsmosca** [comment]
@Krzychumag 

> Krzychumag wrote:
> @ilaripih
> Unfortunately, but not this way. You have to do in the program to make chess engines can read ini file and other files.

Some notes below.
1. Cute Chess is not responsible for reading the ini file of engine.
2. It is the engine that should read its ini file.
3. Example Crafty has ini file crafty.rc. Crafty should read this file and not Cute Chess.
4. If this crafty.rc file is missing, it is not the problem of Cute Chess. It is the problem of the user for failure to put the crafty.rc file in the same directory of the crafty.exe file.
5. To modify the crafty.rc file, it is not the responsibility of Cute Chess. The user should use a notepad.exe (or other text editor), open crafty.rc, modify it and save.

Example egt use in crafty.rc

```
egtb  
tbpath=C:\chess\egt\syzygy
```
6. So that Cute Chess or other GUI, can understand Crafty egtb settings, you may tell Bob (author of Crafty) to support CECP protocol in handling egt.

After receiving the command **protover 2** from gui, Crafty should send to the gui or Cute Chess
`feature egt="syzygy"`

The gui or Cute Chess must understand this, that Crafty is capable of using syzygy egt.
The gui then would build an option for the user to see in the gui for example the following in the Configure Engine\Advanced tab in Cute Chess.
`SyzygyPath: ___________`

The user will fill up the blank with the path of the egt say **c:\chess\egt\syzygy**.

Then the gui will send to the CECP engine:
`egtpath syzygy c:\chess\egt\syzygy`

So now Crafty got the path of the egt of type syzygy (others are nalimov, scorpio) that the user has set in the gui.

This is the best approach for Cute Chess to support egt setting of the engine that is via the CECP protocol, using feature egt="syzygy", and not via crafty.rc file. There is internal communication between the engine and the gui. Engine would send to the gui
`feature egt="syzygy"`
and the gui may understand, that the engine is capable of using syzygy egt.

All CECP engines are encouraged to follow the CECP protocol for setting the engines' egt if those engines support endgame tablebase.

CECP protocol is here.
https://www.gnu.org/software/xboard/engine-intf.html#8

If the CECP engine supports nalimov, the engine should tell the gui.
`feature egt="nalimov" `

Then the gui will tell the engine.
`egtpath nalimov user_nalimov_specified_path`

Also tell Daniel author of ExChess to support the egt feature.

Daniel author of Scorpio had already supported the egt feature using.
`feature egt="scorpio" `

I am not sure if Cute Chess 1.0.0 can understand,
`feature egt="[egt_type]"`
If it understands, it should send to the engine,
`egtpath [egt_type] [path_of_egt]`

**Comment by Krzychumag** [comment]
@fsmosca 
I have in the same directory as the engine Crafty. Unfortunately, but still the program does not see tablebase syzygy.

Here's how I set the Crafty.rc file.

![log](https://user-images.githubusercontent.com/30792281/29530113-a42ed522-86a2-11e7-957e-bc029b04ecba.jpg)
Here is the entry in the Advanced tab in Cute Chess.

![1](https://user-images.githubusercontent.com/30792281/29530263-263e3d50-86a3-11e7-8688-2aeca63048d9.png)

Please write or something bad is still not showing tablebases syzygy.

**Comment by fsmosca** [comment]
Sample setting in my crafty.rc file
```
egtb 
tbpath=C:\myfiles\chess\egt\sygyzy 
```

double clicked the crafty.exe file and I get this

```
unable to open book file [./book.bin].
book is disabled
unable to open book file [./books.bin].

Initializing multiple threads.
System is SMP, not NUMA.
SYZYGY EGTB access enabled, 3 piece TBs found
pondering disabled.
choose from book moves randomly (using weights.)
choose from 4 best moves.
  ..book random set to 1.
book file disabled.

Crafty v25.2

machine has 2 processors

White(1):
```

See that
`SYZYGY EGTB access enabled, 3 piece TBs found`
that means Crafty has read the tbpath.

In Cute Chess it does not show, but the egt path was set correctly in the engine via crafty.rc file. So no problem at all.

Try to double clicked your crafty.exe file, see if it can read the egt path by observing the engine output,
`SYZYGY EGTB access enabled ...`

**Comment by Krzychumag** [comment]
@fsmosca 
I did what you asked. Unfortunately the program does not see tablebases.

![engine](https://user-images.githubusercontent.com/30792281/29531117-63a0b78c-86a7-11e7-9b40-01a10159f7ec.jpg)

**Comment by fsmosca** [comment]
Ahh OK, in that case verify the location of your path.

`D:\chess\egt\syzygy`

Does that path really exist?

There could be problem reading that D drive.
Try to move it to C drive for example.

**Comment by Krzychumag** [comment]
Sorry but I can not because I do not have enough hard disk space.

**Comment by Krzychumag** [comment]
@fsmosca 
And you can show the bookmark " Advanced ".

**Comment by Krzychumag** [comment]
@fsmosca
I have set up the engine this way.

![1](https://user-images.githubusercontent.com/30792281/29532930-d4a1c25c-86af-11e7-8923-dbaea0c97597.jpg)

I ask her why the program does not show in the window with the arrow.

![2](https://user-images.githubusercontent.com/30792281/29532979-09562ccc-86b0-11e7-86d6-39f7768c1cd3.jpg)

I think I'm right and this is a **serious bug** in the program Cute Chess,

**Comment by fsmosca** [comment]
@Krzychumag 
It is good that you successfully set the egt of Crafty via crafty.rc file.

Regarding TB column of Cute Chess, perhaps output of CECP protocol engine is not supported by Cute Chess.

 It also depends how CECP engine would inform the gui or Cute Chess its tbhits. 
For uci engine this is easy since it is in uci protocol like the following example.

`info depth 10 score cp 0 time 100 tbhits 5 pv g1f3`

For CECP engine there is a standard of informing gui or Cute Chess regarding tbhits of engine. But I have doubt if Crafty, or ExChess supported this.

From CECP protocol.

> ply score time nodes pv
> Where:
> 
> ply - Integer giving current search depth.
> score - Integer giving current evaluation in centipawns.
> time - Current search time in centiseconds (ex:1028 = 10.28 seconds).
> nodes Nodes searched.
> *selective depth - Maximium length of any branch in the current search.
> *speed - Nodes per second in last measured time interval.
> *tbhits - Number of tablebase probes made in the current search.
> pv - Freeform text giving current "best" line. You can continue the pv onto another line if you start each continuation line with at least four space characters.
> 

> The items marked with * are optional. If any of these items is present, the pv field must be preceeded directly by a tab character; if no tab character preceeds the first non-integer token, the pv field will start at the first non-blank character after nodes. Otherwise it will start after the last tab that is not behind any non-integer token. Of all integers between nodes and pv the last one is intepreted as tbhits. Of any remaining ones the first is interpreted as selective depth, and a second as speed. More infos could be added to this in the future. Note that older interfaces might consider the optional infos to be part of the pv field, and display them exactly as sent. It is therefore encouraged that engines use tabs or spaces to format this optional info so that it will display nicely in (not too wide) columns.
> 
> A question mark as the last character in the pv field should be used to indicate the reported score is from a fail low, and thus represents an upper bound only. Similarly, an exclamation point should be used to indicate a fail high / lower bound.
> 
> Mate scores should be indicated as 100000 + N for "mate in N moves", and -100000 - N for "mated in N moves".

See section 10 of the CECP protocol.
https://www.gnu.org/software/xboard/engine-intf.html

I think you can create a new issue with regards to showing the tbhits of CECP engine granting that that CECP engine follows the CECP protocol of sending the tbhits info to a gui or Cute Chess.

Based on Crafty output, it does not support sending of tbhits according to CECP protocol.
```
unable to open book file [./book.bin].
book is disabled
unable to open book file [./books.bin].

Initializing multiple threads.
System is SMP, not NUMA.
SYZYGY EGTB access enabled, 3 piece TBs found
pondering disabled.
choose from book moves randomly (using weights.)
choose from 4 best moves.
  ..book random set to 1.
book file disabled.

Crafty v25.2

machine has 2 processors

White(1): xboard

tellicsnoalias set 1 Crafty v25.2 (1 cpus)
tellicsnoalias kibitz Hello from Crafty v25.2! (1 cpus)
protover 2
feature ping=1 setboard=1 san=1 time=1 draw=1
feature sigint=0 sigterm=0 reuse=0 analyze=1
feature myname="Crafty-25.2" name=1
feature playother=1 colors=0 memory=1
feature smp=1
feature variants="normal,nocastle"
feature done=1
post
setboard 2k5/8/8/8/8/4P3/4K3/8 b - - 0 1
st 5
go
       0      1       0 5 1. ... Kd8 2. Kd2 <EGTB>
move Kd8
```

```
0 = ply
1 = score
0 = time
5 = nodes
1. ... Kd8 2. Kd2 <EGTB> = pv
```

So there is no tbhits.

If Crafty wants to support tbhits the CECP protocol way, then it should output the following.

`0      1       0 5 4 1. ... Kd8 2. Kd2 <EGTB>`

That 4 (in this example) between 5 and **1. ... Kd8...** is the tbhits. I don't know if Cute Chess 1.0.0 has supported this type of CECP engine output.

As I understand it, this is not a serious issue, in the first place any CECP engines (in this example Crafty) should at a minimum follows the CECP protocol in sending engine info with tbhits, like the following output.
`[ply] [score] [time] [nodes] [tbhits] [pv]`

Crafty's current output is only,

`[ply] [score] [time] [nodes] [pv]`

So there is no way for any gui or Cute Chess to display the tbhits Crafty.

**Comment by ilaripih** [comment]
> Regarding TB column of Cute Chess, perhaps output of CECP engine is not supported by Cute Chess.

This is true. And I don't even know any Xboard engines that send the `tbhits` info, which is why I left the Xboard implementation for a later time.

**Comment by Krzychumag** [comment]
That's why I write that this is a serious program error Cute Chess. It can not be the case that the program does not show all chess engine options.  And no translation of programmers will be accepted that the program does not support anything. I do not care how chess engine is written, it's important for me to deal with such problems with the program Cute Chess. Unfortunately, the program with such engines can not cope so it is a serious error program Cute Chess.
